### PR TITLE
Add optional `outputSize` to `opts` of `webContents.capturePage([rect, opts])`

### PR DESCRIPTION
**Pull request by Imperat**
#### Description of Change
This PR adds optional parameter `outputSize` to `opts` of `webContents.capturePage([rect, opts])`. The main purpose of this parameter is to improve speed and resources consumption when generating small thumbnails of web pages.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

notes: Added optional `outputSize` to `opts` of `webContents.capturePage([rect, opts])`.

### CODE REVIEW

**Comment by Imperat** [comment]
cc @ckerr @codebytere @zcbenz

File: docs/api/browser-window.md [hunk 1]
```diff
@@ -1162,6 +1162,9 @@ Returns `boolean` - Whether the window's document has been edited.
 * `opts` Object (optional)
   * `stayHidden` boolean (optional) -  Keep the page hidden instead of visible. Default is `false`.
   * `stayAwake` boolean (optional) -  Keep the system awake instead of allowing it to sleep. Default is `false`.
+  * `outputSize` Object (optional)
```
Comment (codebytere) [inline docs/api/browser-window.md:1165]: We should specify the units being used for both width and height.

File: docs/api/web-contents.md [hunk 2]
```diff
@@ -1519,12 +1519,16 @@ console.log(requestId)
 * `opts` Object (optional)
   * `stayHidden` boolean (optional) -  Keep the page hidden instead of visible. Default is `false`.
   * `stayAwake` boolean (optional) -  Keep the system awake instead of allowing it to sleep. Default is `false`.
+  * `outputSize` Object (optional)
```
Comment (codebytere) [inline docs/api/web-contents.md:1522]: Same as above - units need specification!

File: shell/browser/api/electron_api_web_contents.cc [hunk 3]
```diff
@@ -3533,16 +3535,17 @@ v8::Local<v8::Promise> WebContents::CapturePage(gin::Arguments* args) {
...
-  // current system, increase the requested bitmap size to capture it all.
-  gfx::Size bitmap_size = view_size;
-  const gfx::NativeView native_view = view->GetNativeView();
-  const float scale = display::Screen::GetScreen()
-                          ->GetDisplayNearestView(native_view)
-                          .device_scale_factor();
-  if (scale > 1.0f)
-    bitmap_size = gfx::ScaleToCeiledSize(view_size, scale);
+  if (bitmap_size.IsEmpty()) {
+    // By default, the requested bitmap size is the view size in screen
```
Comment (codebytere) [inline shell/browser/api/electron_api_web_contents.cc:3539]: This should also perhaps be specified in the documentation - this might not be assumed behavior by users. Beyond that - is this something we want to do by default? Or is this something that we might want to delegate to users as an option when passing `width` and `height`?

**Review by codebytere — CHANGES_REQUESTED** [review]
Thanks for adding this! I think the performance implications here especially for lower-bandwidth users add a compelling use case for this. We'll definitely also want to add some tests here to verify the new behavior.

**Review by ckerr — COMMENTED** [review]
Code LGTM.

+1 on @codebytere's docs change suggestions

**Comment by Imperat** [comment]
Hi @ckerr @codebytere 
Thanks for reviewing this PR 🙏 
I've added two tests to `spec/api-browser-view-spec.ts` and `spec/api-browser-window-spec.ts`. I was unable to run all of the tests via `npm test` (it takes some time), but I've hardcoded `spec-runner.ts` and verified that `BrowserView.capturePage(rect)` and `BrowserWindow.capturePage(rect)` passed.

I've also updated documentation as you requested 👍 

Please have a look

File: shell/browser/api/electron_api_web_contents.cc (hunk 3 above)
Comment (Imperat) [reply shell/browser/api/electron_api_web_contents.cc:3539]: I've made changes in documentation. 

>Beyond that - is this something we want to do by default?
Not sure, I've made changes in that way to not change or affect existing apps which rely on this API. Currently, Electron does it by default.

> Or is this something that we might want to delegate to users as an option when passing width and height
For the sake of practicality I do not want to add another option to this small PR to control enablement / exposure of scaling feature. However, I am happy to raise a separate PR to introduce this option once the current PR reaches `main` :)

File: docs/api/browser-window.md (hunk 1 above)
Comment (Imperat) [reply docs/api/browser-window.md]: added: _The desired width and height in pixels (positive numbers) of the returned image_

File: docs/api/web-contents.md (hunk 2 above)
Comment (Imperat) [reply docs/api/web-contents.md]: added: _The desired width and height in pixels (positive numbers) of the returned image_

File: shell/browser/api/electron_api_web_contents.cc (hunk 3 above)
Comment (codebytere) [reply shell/browser/api/electron_api_web_contents.cc:3539]: Fair enough! I don't feel strongly about it really :)

File: docs/api/web-contents.md [hunk 4]
```diff
@@ -1519,13 +1519,19 @@ console.log(requestId)
...
+  * `outputSize` [Size](structures/size.md) (optional) - The desired width and height in pixels (positive numbers) of the returned image.
 
 Returns `Promise<NativeImage>` - Resolves with a [NativeImage](native-image.md)
 
 Captures a snapshot of the page within `rect`. Omitting `rect` will capture the whole visible page.
 The page is considered visible when its browser window is hidden and the capturer count is non-zero.
 If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
 
+The size of returned [NativeImage](native-image.md) will be `outputSize` if it was provided.
+Otherwise, the size would be large enought to fit all of the available pixels in the current system.
```
Comment (codebytere) [inline docs/api/web-contents.md:1531]: We should validate this with a test! Ex. validate that if the default size would be `{ 200, 200 }` and i set outputSize to `{ 1500, 150 }` the resultant image should have a size of `{ 1500, 150 }`.

File: docs/api/browser-window.md [hunk 5]
```diff
@@ -1162,6 +1162,18 @@ Returns `boolean` - Whether the window's document has been edited.
 * `opts` Object (optional)
   * `stayHidden` boolean (optional) -  Keep the page hidden instead of visible. Default is `false`.
   * `stayAwake` boolean (optional) -  Keep the system awake instead of allowing it to sleep. Default is `false`.
+  * `outputSize` [Size](structures/size.md) (optional) - The desired width and height in pixels (positive numbers) of the returned image.
+
+Returns `Promise<NativeImage>` - Resolves with a [NativeImage](native-image.md)
+
+Captures a snapshot of the page within `rect`. Omitting `rect` will capture the whole visible page.
+The page is considered visible when its browser window is hidden and the capturer count is non-zero.
+If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
```
Comment (codebytere) [inline docs/api/browser-window.md:1171]: Looks like this got duplicated? It's [here](https://github.com/electron/electron/pull/41281/files#diff-e8a455d5c2bb8c09554011e09a285d6e66c3c681082ae9998f2a4a026eaf0b03R1178) already.

**Review by codebytere — COMMENTED** [review]
Thanks for addressing this - once there's one more test for large edge size validation and the duplicated docs get removed i think we can get this moving!

File: docs/api/browser-window.md (hunk 5 above)
Comment (Imperat) [reply docs/api/browser-window.md]: Sorry, I missed that. [Fixed in the last commit](https://github.com/electron/electron/pull/41281/commits/861d79b9f12cb9b09055d2580925203110e54bb1)

File: docs/api/web-contents.md (hunk 4 above)
Comment (Imperat) [reply docs/api/web-contents.md:1531]: Good suggestion!
I've elaborated the test into three scenarios: 
1. requested size is smaller than captured area
2. one dimension of requested size is bigger than captured area
3. both dimensions are bigger than captured area

**Review by codebytere — APPROVED** [review]
API LGTM

File: docs/api/web-contents.md (hunk 4 above)
Comment (jkleinsc) [inline docs/api/web-contents.md:1531]: Same comment above. - this statement is confusing as is.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
The documentation needs clarity added around the description of what happens when outputSize is not provided.

File: docs/api/browser-window.md [hunk 6]
```diff
@@ -1154,10 +1154,18 @@ Returns `boolean` - Whether the window's document has been edited.
...
 
 Returns `Promise<NativeImage>` - Resolves with a [NativeImage](native-image.md)
 
-Captures a snapshot of the page within `rect`. Omitting `rect` will capture the whole visible page. If the page is not visible, `rect` may be empty. The page is considered visible when its browser window is hidden and the capturer count is non-zero. If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
+Captures a snapshot of the page within `rect`. Omitting `rect` will capture the whole visible page.
+If the page is not visible, `rect` may be empty. The page is considered visible when its browser window is hidden and the capturer count is non-zero.
+If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
+
+The size of returned [NativeImage](native-image.md) will be `outputSize` if it was provided.
+Otherwise, the size would be large enought to fit all of the available pixels in the current system.
```
Comment (nornagon) [inline docs/api/browser-window.md:1166]: This is confusing, what does "all of the available pixels in the current system" mean?

How does this behave if the aspect ratio of `outputSize` differs from the aspect ratio of `rect`?

Comment (Imperat) [reply docs/api/browser-window.md]: For "all of the available pixels":
Initially, I took this phrase somewhere from Chromium sources, but I agree it sounds confusing.

I rephrased it to state the default value instead:

> The desired width and height in pixels (positive numbers) of the returned image. Default is the size of WebContents in physical pixels, that is WebContents height and WebContents width multiplied by device pixel ratio.

For aspect ratio:

I've added a note:
> **Note:** When using `outputSize` it is responsibility of caller to preserve aspect ratio. If aspect ratio of `outputSize` is not the same as page aspect ratio the output image will be scaled.

**Comment by Imperat** [comment]
Hi @jkleinsc @nornagon 

Thank you for your feedback. I apologize for not being able to update the PR earlier when I received your review.
I made some changes now. Could you please have a second look? 🙏

File: spec/api-browser-window-spec.ts [hunk 7]
```diff
@@ -2307,6 +2307,76 @@ describe('BrowserWindow module', () => {
...
+              height: 1500
+            }
+          });
+
+          expect(image.getSize()).to.deep.equal({ width: 150, height: 1500 });
+        });
+      });
+
+      describe('when both dimensions of `outputSize` are bigger than both dimensions of the captured area', () => {
+        it('returns image with requested size', async () => {
```
Comment (samuelmaddock) [inline spec/api-browser-window-spec.ts:2357]: It's fine as is, but this should probably just use one `it` like
```suggestion
      it('returns image with requested size when both dimensions of `outputSize` are bigger than both dimensions of the captured area', async () => {
```

**Review by samuelmaddock — APPROVED** [review]
API LGTM

**Comment by samuelmaddock** [comment]
@Imperat your PR might need to be rebased to resolve the build issues being seen.

**Comment by samuelmaddock** [comment]
I'm adding a wip label on this PR until the requested changes can be implemented. Please tag `@electron/wg-api` once it's ready to be reviewed again and we'll take another look! 🙇

**Comment by ckerr** [comment]
This PR is quite old; is this something we want to revisit or close?

@jkleinsc did the followup revisions meet your change request?

**Comment by codebytere** [comment]
@Imperat could you rebase this?

**Review by jkleinsc — APPROVED** [review]

File: spec/api-browser-view-spec.ts [hunk 8]
```diff
@@ -779,6 +779,96 @@ describe('BrowserView module', () => {
...
+          y: 0,
+          width: 100,
+          height: 100
+        }, {
+          outputSize: {
+            width: 1500,
+            height: 150
+          }
+        });
+        expect(image.getSize().width).to.be.lessThan(1500);
```
Comment (codebytere) [inline spec/api-browser-view-spec.ts]: We should check the whole image, as well as checking that it's limited to default size as per docs - 

```suggestion
        expect(image.getSize()).to.deep.equal({ width: 200, height: 150 });
```

File: spec/api-browser-view-spec.ts [hunk 9]
```diff
@@ -779,6 +779,96 @@ describe('BrowserView module', () => {
...
+          width: 100,
+          height: 100
+        }, {
+          outputSize: {
+            width: 1500,
+            height: 1500
+          }
+        });
+        expect(image.getSize().width).to.be.lessThan(1500);
+        expect(image.getSize().height).to.be.lessThan(1500);
```
Comment (codebytere) [inline spec/api-browser-view-spec.ts]: Same as above:

```suggestion
        expect(image.getSize()).to.deep.equal({ width: 200, height: 200 });
```

File: spec/api-browser-window-spec.ts [hunk 10]
```diff
@@ -2434,6 +2434,70 @@ describe('BrowserWindow module', () => {
...
+          y: 0,
+          width: 100,
+          height: 100
+        }, {
+          outputSize: {
+            width: 150,
+            height: 1500
+          }
+        });
+        expect(image.getSize().height).to.be.lessThan(1500);
```
Comment (codebytere) [inline spec/api-browser-window-spec.ts]: expect(image.getSize()).to.deep.equal({ width: 150, height: 200 });

File: spec/api-browser-window-spec.ts [hunk 11]
```diff
@@ -2434,6 +2434,70 @@ describe('BrowserWindow module', () => {
...
+          height: 100
+        }, {
+          outputSize: {
+            width: 1500,
+            height: 1500
+          }
+        });
+
+        expect(image.getSize().width).to.be.lessThan(1500);
+        expect(image.getSize().height).to.be.lessThan(1500);
```
Comment (codebytere) [inline spec/api-browser-window-spec.ts]:
```suggestion
        expect(image.getSize()).to.deep.equal({ width: 200, height: 200 });
```

**Comment by jkleinsc** [comment]
I was hoping to revive this PR for merging, but it appears that this change causes a sporadic `UnknownVizError` when calling `view->CopyFromSurface` with sizes other than the original size, eg see https://github.com/electron/electron/actions/runs/21637443235/job/62370575614?pr=41281#step:21:263; https://github.com/electron/electron/actions/runs/21404630755/job/61668994107#step:20:459

Given this, I do not believe this PR as it stands should be merged.   Additionally, the functionality proposed in this PR can instead be accomplished by existing API calls:
```js
const image = await w.capturePage({
  x: 0,
  y: 0,
  width: 100,
  height: 100
});
const smallerImage = image.resize({
  width: 20,
  height: 20
});
```

I am going to close this PR as it stands right now.  @Imperat if you want to pursue this further, the sporadic `UnknownVizError` issue would need to be resolved, but the API would not need additional approval.

**Comment by Imperat** [comment]
Hi @jkleinsc 

Thank you for your effort to attempt to merge this PR 🙏 
I will try to return to it once I have some free time
